File: selection/_007/Main.java
package selection._007;

import java.util.ArrayList;
import java.util.Scanner;
import java.util.TreeMap;
import java.util.Map.Entry;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        TreeMap<Integer, ArrayList<Integer>> map = new TreeMap<>();
        for (int i = 0; i < n; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            map.computeIfAbsent(x, k -> new ArrayList<>()).add(y);
        }
        sc.close();

        int result = 0;
        for (Entry<Integer, ArrayList<Integer>> entry1 : map.entrySet()) {
            for (Entry<Integer, ArrayList<Integer>> entry2 : map.entrySet()) {
                for (Integer y1 : entry1.getValue()) {
                    for (Integer y2 : entry2.getValue()) {
                        int x1 = entry1.getKey();
                        int x2 = entry2.getKey();
                        if (x1 < x2) {
                            if (map.containsKey(x2 - y2 + y1) && map.get(x2 - y2 + y1).contains(y2 + x2 - x1)
                                    && map.containsKey(
                                            x1 - y2 + y1)
                                    && map.get(x1 - y2 + y1).contains(y1 + x2 - x1)) {
                                int diffX = Math.abs(x2 - x1);
                                int diffY = Math.abs(y2 - y1);
                                result = Math.max(result, diffX * diffX + diffY * diffY);
                            }
                        }
                    }
                }
            }
        }

        System.out.println(result);
    }
}

File: selection/_007/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	var n int
	fmt.Scan(&n)

	m := make([][]bool, 5001)
	for i := 0; i <= 5000; i++ {
		m[i] = make([]bool, 5001)
	}

	x := make([]int, n)
	y := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&x[i], &y[i])
		m[x[i]][y[i]] = true
	}

	result := 0
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			// 反時計に90°回転
			nx := -y[j] + y[i] + x[i]
			ny := x[j] - x[i] + y[i]
			if nx < 0 || nx > 5000 || ny < 0 || ny > 5000 {
				continue
			}
			if !m[nx][ny] {
				continue
			}

			nx2 := x[j] - y[j] + y[i]
			ny2 := y[j] + x[j] - x[i]
			if nx2 < 0 || nx2 > 5000 || ny2 < 0 || ny2 > 5000 {
				continue
			}
			if !m[nx2][ny2] {
				continue
			}

			result = int(math.Max(float64(result), math.Pow(float64(x[i]-x[j]), 2)+math.Pow(float64(y[i]-y[j]), 2)))
		}
	}

	fmt.Println(result)
}
